core/execution/engine.go

```diff
@@ -614,17 +617,93 @@ func (e *Engine) SubmitStopOrders(
 	idgen common.IDGenerator,
 	fallsBelowID *string,
 	risesAboveID *string,
-) error {
-	return errors.New("stop order submission not supported yet")
+) (*types.OrderConfirmation, error) {
+	var market string
+	if submission.FallsBelow != nil {
+		market = submission.FallsBelow.OrderSubmission.MarketID
+	} else {
+		market = submission.RisesAbove.OrderSubmission.MarketID
+	}
+
+	mkt, ok := e.markets[market]
+	if !ok {
+		return nil, types.ErrInvalidMarketID
+	}
+
+	conf, err := mkt.SubmitStopOrdersWithIDGeneratorAndOrderIDs(
+		ctx, submission, party, idgen, fallsBelowID, risesAboveID)
+	if err != nil {
+		return nil, err
+	}
+
+	// not necessary going to trade on submission, could be nil
+	if conf != nil {
+		// increasing the gauge, just because we reuse the
+		// decrement function, and it required the order + passive
+		metrics.OrderGaugeAdd(1, market)
+		e.decrementOrderGaugeMetrics(market, conf.Order, conf.PassiveOrdersAffected)
+	}
+
+	return conf, nil
 }
 
 func (e *Engine) CancelStopOrders(
 	ctx context.Context,
-	cancellation *types.StopOrdersCancellation,
+	cancel *types.StopOrdersCancellation,
 	party string,
 	idgen common.IDGenerator,
 ) error {
-	return errors.New("stop order cancellation not supported yet")
+	// ensure that if orderID is specified marketId is as well
+	if len(cancel.OrderID) > 0 && len(cancel.MarketID) <= 0 {
+		return ErrInvalidOrderCancellation
+	}
+
+	if len(cancel.MarketID) > 0 {
+		if len(cancel.OrderID) > 0 {
+			return e.cancelStopOrders(ctx, party, cancel.MarketID, cancel.OrderID, idgen)
+		}
+		return e.cancelStopOrdersByMarket(ctx, party, cancel.MarketID)
+	}
+	return e.cancelAllPartyStopOrders(ctx, party)
```

Comment:
This is a bit odd: I can cancel all stop orders, across all markets, regardless of asset? I mean: it's a valid thing to do, but it looks like that'd be essentially sending an empty `StopOrdersCancellation{}` message. Feels a bit risky, and it may result in people wanting to be able to cancel stop orders on a per-asset basis or something (which we could implement, albeit in a bit of a hacky way)

Reply:
We allow the same behaviour to orders to be fair. So just went for the same thing